File: ctci/ch01/cpp/urlify.cpp
#include <iostream>
#include <sstream>
#include <string>

using namespace std;

string urlify(string str, int length) {
  string result = "";

  int count = 0;
  bool startCount = false;

  for (char c : str) {
    if (c == ' ') {
      if (!startCount)
        continue;
      if (count >= length)
        break;
      count += 1;
      result += "%20";
    } else {
      startCount = true;
      count += 1;
      result += c;
    }
  }

  return result;
}

int main(int argc, char *argv[]) {
  if (argc != 3)
    return 1;

  istringstream ss(argv[2]);
  int length;
  if (!(ss >> length)) {
    cerr << "invalid number: " << argv[2] << endl;
  } else if (!ss.eof()) {
    cerr << "trailing character: " << argv[2] << endl;
  }

  cout << urlify(argv[1], length) << endl;
}

File: ctci/ch01/js/urlify.ts
export function urlify(str: string, length: number): string {
  let url = ""
  if (length == 0) return url;

  let startCounting = false;
  let count = 0;

  for (let i = 0; i < str.length; i++) {
    if (str[i] == " ") {
      if (!startCounting) continue;
      if (count >= length) break;
      count += 1;
      url += "%20";
    } else {
      startCounting = true;
      count += 1;
      url += str[i];
    }
  }
  return url;
}

export function urlifyConvenient(str: string, length: number): string {
  return str.trim().replaceAll(" ", "%20");
}
